### ISSUE TITLE
[question] conan graph info with multiple remotes

### ISSUE THREAD

**Issue by nextsilicon-itay-bookstein** [conan-io/conan#16791]
### What is your question?

Hi!

As part of our transition to Conan 2 (specifically 2.6), I'm trying to stand up a local artifactory repository so that we can have binaries ready for developers with all our profiles. I add it as a remote with index=0. I have a script that's supposed to build the dependency graph for a `conanfile.txt`. When that script runs `conan graph info --update` (with both our remote and conancenter enabled), for all immediate dependencies in the `conanfile.txt`, it checks both remotes, and downloads recipes from conancenter as ours is still empty. However, when it tries to resolve indirect dependencies, it fails on getting a 404 from our remote:

```
ERROR: Package 'c-ares/[>=1.19.1 <2]' not resolved: {
  "errors" : [ {
    "status" : 404,
    "message" : "File not found."
  } ]
}. [Remote: myremote].
```

I'm specifically trying to avoid forking conan-center-index for now, and standing up a CI flow that will push recipes to our private repository. I might eventually do it later. So the questions are as follows:
1. Is this behavior expected?
2. Is there a better way to achieve what I'm trying to do?

Thanks!
Itay

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by AbrilRBS** [comment]
Hi @nextsilicon-itay-bookstein thanks a lot for taking the time to report your issue.

So far I have not been able to reproduce your 404 issue, and I can't find the reason for the behaviour, but I'm sure someone else will chime in with the answer I'm missing :)

I did want to talk about

> 2. Is there a better way to achieve what I'm trying to do?

Have you taken a look into [the `conan graph build-order` command](https://docs.conan.io/2/reference/commands/graph/build_order.html)? It will help you get the information you need to properly build each of your packages and its (transitive) dependencies in the correct order, which will be easier than trying to parse `conan graph info` output 1 by 1 - Let me know if you need help in that regard, happy to help where possible!

**Comment by nextsilicon-itay-bookstein** [comment]
Ah, yes, I'm running `conan graph build-order` immediately afterwards, and parallelizing within the graph levels it produces.
I'm using `conan graph info` to produce a graph I could convert to a pkglist that I could later `conan upload` against, because as far as I can tell `conan graph build-order` has no option to produce a regular graph usable for converting into a pkglist. The reason I want the graph's pkglist rather than the install commands' pkglists is that I thought it would be more convenient if my automation were able to upload cached-but-not-uploaded-yet packages.

I'll see if I can produce an easy repro without all the custom stuff I'm doing. Maybe if I'm not able to it will point towards something I'm doing wrong.

**Comment by nextsilicon-itay-bookstein** [comment]
So, as far as reproducing this issue:
```
> conan profile detect
> conan remote add --index=0 myremote <url>
> conan graph info --require=grpc/1.54.3

======== Computing dependency graph ========
grpc/1.54.3: Not found in local cache, looking in remotes...
grpc/1.54.3: Checking remote: myremote
grpc/1.54.3: Checking remote: conancenter
grpc/1.54.3: Downloaded recipe revision f477bbbb4b6464d8add7b40c4b5007d7
Graph root
    cli
Requirements
    grpc/1.54.3#f477bbbb4b6464d8add7b40c4b5007d7 - Downloaded (conancenter)

-------- Basic graph information --------
... <output about grpc> ...
ERROR: Package 'abseil/[>=20230125.3 <=20230802.1]' not resolved: {
  "errors" : [ {
    "status" : 404,
    "message" : "File not found."
  } ]
}. [Remote: myremote].
```
These commands reproduce the issue for me, with the implicit assumption that `myremote` is completely empty.
You'll note that the grpc recipe is looked for in both remotes, but attempting to resolve its abseil requirement fails immediately on the first remote (with index=0).

**Comment by nextsilicon-itay-bookstein** [comment]
Upon debugging the discrepancy, the direct requirement resolves through `conans.client.graph.proxy.ConanProxy._find_newest_recipe_in_remotes`, which loops over remotes and catches `NotFoundException`, while the indirect requirement resolves through `conans.client.graph.range_resolver.RangeResolver._resolve_remote`, which loops over remotes but **does not** catch `NotFoundException`.

**Comment by memsharded** [comment]
I have tried to reproduce that setup, but still didn't manage to make it fail.

I think it might be related to permissions, the 404 from some servers is a specific configuration that returns a 404 for any request if user is not authenticated, instead of throwing an auth challenge to get the credentials or fail with a 401 or 403 instead. I have done some quick attempts, but I haven't been able to reproduce it this way either.

**Comment by memsharded** [comment]
Some information about the server might be useful too, I don't think it is related, but just in case.

**Comment by nextsilicon-itay-bookstein** [comment]
After some rummaging around, I found the issue: it was that I mis-specified the URL. I used the value of the "URL to file:" field from the artifactory UI. Only after finding [this](https://docs.conan.io/en/1.6/uploading_packages/artifactory_ce.html) page did I understand how to derive the API URL from the name of the repository that I configured. The wrong URL always returned 404, injecting faults in places where a properly configured remote wouldn't.
